frontend/src/__tests__/unit/useGroups.spec.ts

```diff
@@ -0,0 +1,123 @@
+    const error = {
+      statusObject: {
+        code: 403,
+      },
+    };
+    k8sListResourceMock.mockReturnValue(Promise.reject(error));
+
+    const renderResult = testHook(useGroups);
+    expect(k8sListResourceMock).toHaveBeenCalledTimes(1);
+    expectHook(renderResult)
+      .toStrictEqual([[], false, undefined, expect.any(Function)])
+      .toHaveUpdateCount(1);
+
+    // wait for update
+    await renderResult.waitForNextUpdate();
+    expect(k8sListResourceMock).toHaveBeenCalledTimes(1);
+    expectHook(renderResult)
+      .toStrictEqual([[], true, undefined, expect.any(Function)])
+      .toHaveUpdateCount(2)
+      .toBeStable([false, false, true, true]);
+
+    // refresh
+    await act(() => renderResult.result.current[3]());
+    // error 403 should cache error and prevent subsequent attempts to fetch
+    expect(k8sListResourceMock).toHaveBeenCalledTimes(1);
+    expectHook(renderResult).toHaveUpdateCount(3).toBeStable([true, true, true, true]);
+  });
+
+  it('should handle 404 as an error', async () => {
+    const error = {
+      statusObject: {
+        code: 404,
+      },
+    };
+    k8sListResourceMock.mockReturnValue(Promise.reject(error));
+
+    const renderResult = testHook(useGroups);
+
+    expect(k8sListResourceMock).toHaveBeenCalledTimes(1);
+    expectHook(renderResult)
+      .toStrictEqual([[], false, undefined, expect.any(Function)])
+      .toHaveUpdateCount(1);
+
+    // wait for update
+    await renderResult.waitForNextUpdate();
+    expectHook(renderResult)
+      .toStrictEqual([[], false, new Error('No groups found.'), expect.any(Function)])
+      .toHaveUpdateCount(2)
+      .toBeStable([true, true, false, true]);
+
+    expect(k8sListResourceMock).toHaveBeenCalledTimes(1);
+
+    // refresh
+    await act(() => renderResult.result.current[3]());
+    expect(k8sListResourceMock).toHaveBeenCalledTimes(2);
+    // we get a new error because the k8s API is called a 2nd time
+    expectHook(renderResult).toHaveUpdateCount(3).toBeStable([true, true, false, true]);
+  });
+
+  it('should handle other errors and rethrow', async () => {
+    const error = new Error();
+    k8sListResourceMock.mockReturnValue(Promise.reject(new Error()));
+
+    const renderResult = testHook(useGroups);
+    expect(k8sListResourceMock).toHaveBeenCalledTimes(1);
+    expectHook(renderResult)
+      .toStrictEqual([[], false, undefined, expect.any(Function)])
+      .toHaveUpdateCount(1);
+
+    // wait for update
+    await renderResult.waitForNextUpdate();
+    expect(k8sListResourceMock).toHaveBeenCalledTimes(1);
+    expectHook(renderResult)
+      .toStrictEqual([[], false, error, expect.any(Function)])
+      .toHaveUpdateCount(2)
+      .toBeStable([true, true, false, true]);
+
+    // refresh
+    await act(() => renderResult.result.current[3]());
+    expect(k8sListResourceMock).toHaveBeenCalledTimes(2);
+    expectHook(renderResult).toHaveUpdateCount(3).toBeStable([true, true, true, true]);
```

Comment:
```suggestion
    expectHook(renderResult).toHaveUpdateCount(3).toBeStable([false, true, true, true]);
```
To my other comment, this should change or we need to implement something in addition to `NotReadyYet` error and keep the default. Which is something we could consider `new NothingToChange()` sorta deal

Reply:
I will update the test to throw a new error instead of reusing the same error. This would align with real world error scenarios. Therefore the error value will not be stable, however the default value will remain stable.

Reply:
Right, if the error updates -- the data stays the same ([I believe these are the scenarios](https://github.com/opendatahub-io/odh-dashboard/pull/1630#discussion_r1297397396))